pallets/collator-assignment/src/lib.rs

```diff
@@ -159,16 +169,110 @@ pub mod pallet {
+            // We need to sort both separately as we have fixed space for parachains at the moment
+            // which means even when we have some parathread cores empty we cannot schedule parachain there.
+            // We need to change this once other part of algorithm start to differentiate between
+            // bulk and pool paras while deducting the tip amount.
+            if !enough_cores_for_bulk_paras {
+                bulk_paras.sort_by(|a, b| {
+                    T::CollatorAssignmentTip::get_para_tip(b.para_id)
+                        .cmp(&T::CollatorAssignmentTip::get_para_tip(a.para_id))
+                });
+
+                pool_paras.sort_by(|a, b| {
+                    T::CollatorAssignmentTip::get_para_tip(b.para_id)
+                        .cmp(&T::CollatorAssignmentTip::get_para_tip(a.para_id))
+                });
+            }
+
+            bulk_paras.truncate(max_number_of_bulk_paras as usize);
+            // We are not truncating pool paras, since their workload is not continuous one core
+            // can be shared by many paras during the session.
+
+            let (ordered_chains, sorted_chains_based_on_tip) = Self::order_paras(
+                bulk_paras,
+                pool_paras,
+                target_session_index,
+                number_of_collators,
+                collators_per_container,
+                collators_per_parathread,
+            );
+
+            // We should charge tip if either this method or the order_para method has sorted chains based on tip
+            // So in other words: if parachain demand exceeds the `max_number_of_bulk_paras` OR
+            // if `num_collators` is not enough to satisfy  collation need of all paras.
+            (
+                ordered_chains,
+                sorted_chains_based_on_tip || !enough_cores_for_bulk_paras,
+            )
+        }
+
+        pub(crate) fn order_paras(
+            mut bulk_paras: Vec<ChainNumCollators>,
+            pool_paras: Vec<ChainNumCollators>,
+            target_session_index: T::SessionIndex,
+            number_of_collators: u32,
+            collators_per_container: u32,
+            collators_per_parathread: u32,
+        ) -> (Vec<ChainNumCollators>, bool) {
+            // Are there enough collators to satisfy the minimum demand?
+            let enough_collators_for_all_chain = number_of_collators
+                >= T::HostConfiguration::min_collators_for_orchestrator(target_session_index)
+                    .saturating_add(collators_per_container.saturating_mul(bulk_paras.len() as u32))
+                    .saturating_add(
+                        collators_per_parathread.saturating_mul(pool_paras.len() as u32),
+                    );
+
+            let mut chains: Vec<_> = bulk_paras.drain(..).chain(pool_paras).collect();
+
+            // Prioritize paras by tip on congestion
+            // As of now this doesn't distinguish between bulk paras and pool paras
```

Comment:
So if all parathreads have a high fee and parachains have low fee, it is possible for collators to be assigned only to parathreads?

Reply:
Yes. This is original behavior for dance box which I have not touched as it is out of scope.

Reply:
*disregard*